`test_kafka_bad_messages/test_1.py::test_log_to_exceptions` is flaky again (librdkafka log-order race)

## Description

`test_kafka_bad_messages/test_1.py::test_log_to_exceptions` is flaky on the `Integration tests (amd_msan, 2/6)` build. CIDB shows 11 hits on 2026-05-19 alone (3 on master, 8 across unrelated PRs), plus 5 hits on 2026-05-18 and 1 on 2026-04-24.

This is a regression of the same test class as #83694, but with a different root cause. The original #83694 (`Bind for 0.0.0.0:2181 failed: port is already allocated`) was a Docker port conflict, fixed by backporting `#81108` to 25.5 and closed 2025-07-24. The current failure is a log-ordering race introduced when `librdkafka` was bumped to 2.14.1 in `#105222`.

## Failure mode

`test_log_to_exceptions` points a `Kafka`-engine table at `localhost:9876` (an unused port), runs a `SELECT`, then asserts:

```python
system_kafka_consumers_content = instance.query(
    "SELECT exceptions.text FROM system.kafka_consumers ARRAY JOIN exceptions WHERE table LIKE 'foo_exceptions' LIMIT 1"
)
assert system_kafka_consumers_content.startswith(
    f"[thrd:localhost:{non_existent_broker_port}/bootstrap]: 1/1 brokers are down"
)
```

`librdkafka` emits two log lines for an unreachable broker:

1. `[thrd:localhost:9876/bootstrap]: localhost:9876/bootstrap: Connect to ipv4#127.0.0.1:9876 failed: Connection refused (after 0ms in state CONNECT)` (per-attempt, from `rd_kafka_broker_fail`)
2. `[thrd:localhost:9876/bootstrap]: 1/1 brokers are down` (state-change summary)

Both go through `KafkaConfigLoader::log_callback_proxy` -> `KafkaConsumer::setExceptionInfo` and end up in `system.kafka_consumers.exceptions`. Their order inside that array is not deterministic; it depends on `librdkafka` broker-thread timing, and under MSAN (slow execution) the `Connect ... failed: Connection refused` line frequently lands at index 0, while the test asserts the `1/1 brokers are down` line is first.

The `ebeda53c2555` (in `#105222`) test update only switched which of the two messages the assertion expected to be first, so the test stayed flaky in the other direction. CIDB confirms both variants flake in the same 30-day window:

```text
expected log prefix                                                    failures
[thrd:localhost:9876/bootstrap]: 1/1 brokers are down                  11
[thrd:localhost:9876/bootstrap]: ...Connect to ...: Connection refused  5
```

## Reproducer

```text
File: test_kafka_bad_messages/test_1.py:115 - in test_log_to_exceptions
    assert system_kafka_consumers_content.startswith(
E   assert False
E    +  where False = <built-in method startswith of str ...>('[thrd:localhost:9876/bootstrap]: 1/1 brokers are down')
E    +    where ... = "[thrd:localhost:9876/bootstrap]: localhost:9876/bootstrap: Connect to ipv4#127.0.0.1:9876 failed: Connection refused (after 0ms in state CONNECT)\n0. src/Common/StackTrace.cpp:512:12: StackTrace::StackTrace() ...".startswith
```

Example CI run: https://s3.amazonaws.com/clickhouse-test-reports/json.html?REF=master&sha=916ce6133af5bd2f95cedeb67f0ffe88bfb327d7&name_0=MasterCI&name_1=Integration%20tests%20%28amd_msan%2C%202%2F6%29

## CIDB evidence (last 30 days)

```text
day         total  master_hits  pr_hits  distinct_prs
2026-05-19   11     3            8        9
2026-05-18    5     0            5        1   (all PR #105222 librdkafka bump)
2026-04-24    1     0            1        1
```

## Fix proposal

Remove the order dependency in the assertion: query `system.kafka_consumers.exceptions` for any entry that `startsWith` the expected `1/1 brokers are down` log line, and use `query_with_retry` to wait for `librdkafka` to emit it (which it always eventually does, regardless of which log message lands first). The test still verifies the same specific log line is captured by the consumer; it just no longer relies on the unspecified order of `librdkafka` broker-thread events.

The fix is a test-only change, no `librdkafka` or `Kafka` engine code is modified.

## Context

- Original issue: #83694 (closed 2025-07-24, different root cause: Docker port conflict)
- `librdkafka` bump that exposed the new race: #105222 (merged 2026-05-18)
- Resurfacing reported by @alexey-milovidov on #102866 (2026-05-20)

<!-- ch-version-info:start -->
### Version info
- Resolved by: #105395
- Merged into: `26.5.1.899` (included in `26.5` and later)
- Backported to: `26.5.6.81`, `26.4.5.35`, `26.3.17.81`
<!-- ch-version-info:end -->

--- a/tests/integration/test_kafka_bad_messages/test_1.py
+++ b/tests/integration/test_kafka_bad_messages/test_1.py
@@ -107,14 +107,19 @@ def test_log_to_exceptions(kafka_cluster, max_retries=20):
     )
     instance.query("SYSTEM FLUSH LOGS")
 
-    # `librdkafka` emits several log lines when the broker is unreachable
-    # (per-attempt `Connect ... failed: Connection refused` plus a periodic
-    # `N/M brokers are down` summary). Their order inside
-    # `system.kafka_consumers.exceptions` is not guaranteed and depends on
-    # broker-thread timing, so look for the expected log line in any position
-    # and retry until it has been surfaced.
-    expected_prefix = (
-        f"[thrd:localhost:{non_existent_broker_port}/bootstrap]: 1/1 brokers are down"
+    # `librdkafka` emits several log lines when the broker is unreachable.
+    # Both flow into `system.kafka_consumers.exceptions` via
+    # `KafkaConsumer::setExceptionInfo`:
+    #   - per-attempt `Connect to ... failed: Connection refused`
+    #   - periodic `N/M brokers are down` summary
+    # There is no guarantee that any specific line shows up first or at all
+    # under unusual broker-thread timing, so accept either form.
+    thrd_prefix = (
+        f"[thrd:localhost:{non_existent_broker_port}/bootstrap]:"
+    )
+    broker_down_marker = f"{thrd_prefix} 1/1 brokers are down"
+    connect_refused_marker = (
+        f"{thrd_prefix} localhost:{non_existent_broker_port}/bootstrap: Connect to"
     )
     matching_count = instance.query_with_retry(
         f"""
@@ -122,7 +127,8 @@ def test_log_to_exceptions(kafka_cluster, max_retries=20):
         FROM system.kafka_consumers
         ARRAY JOIN exceptions
         WHERE table = 'foo_exceptions'
-          AND startsWith(exceptions.text, '{expected_prefix}')
+          AND (startsWith(exceptions.text, '{broker_down_marker}')
+               OR startsWith(exceptions.text, '{connect_refused_marker}'))
         """,
         check_callback=lambda res: int(res.strip()) >= 1,
         retry_count=max_retries,
@@ -136,7 +142,8 @@ def test_log_to_exceptions(kafka_cluster, max_retries=20):
         logging.debug(f"system.kafka_consumers content: {all_exceptions}")
         raise AssertionError(
             f"Expected at least one entry in system.kafka_consumers.exceptions starting with "
-            f"{expected_prefix!r}, but none was found. Captured exceptions:\n{all_exceptions}"
+            f"either {broker_down_marker!r} or {connect_refused_marker!r}, "
+            f"but none was found. Captured exceptions:\n{all_exceptions}"
         )
 
     instance.query("DROP TABLE foo_exceptions")

--- a/tests/integration/test_kafka_bad_messages/test_1.py
+++ b/tests/integration/test_kafka_bad_messages/test_1.py
@@ -107,14 +107,37 @@ def test_log_to_exceptions(kafka_cluster, max_retries=20):
     )
     instance.query("SYSTEM FLUSH LOGS")
 
-    system_kafka_consumers_content = instance.query(
-        "SELECT exceptions.text FROM system.kafka_consumers ARRAY JOIN exceptions WHERE table LIKE 'foo_exceptions' LIMIT 1"
-    )
-
-    logging.debug(f"system.kafka_consumers content: {system_kafka_consumers_content}")
-    assert system_kafka_consumers_content.startswith(
+    # `librdkafka` emits several log lines when the broker is unreachable
+    # (per-attempt `Connect ... failed: Connection refused` plus a periodic
+    # `N/M brokers are down` summary). Their order inside
+    # `system.kafka_consumers.exceptions` is not guaranteed and depends on
+    # broker-thread timing, so look for the expected log line in any position
+    # and retry until it has been surfaced.
+    expected_prefix = (
         f"[thrd:localhost:{non_existent_broker_port}/bootstrap]: 1/1 brokers are down"
     )
+    matching_count = instance.query_with_retry(
+        f"""
+        SELECT count()
+        FROM system.kafka_consumers
+        ARRAY JOIN exceptions
+        WHERE table = 'foo_exceptions'
+          AND startsWith(exceptions.text, '{expected_prefix}')
+        """,
+        check_callback=lambda res: int(res.strip()) >= 1,
+        retry_count=max_retries,
+        sleep_time=1,
+    )
+    if int(matching_count.strip()) < 1:
+        # Surface the full exceptions array on failure to make debugging easier.
+        all_exceptions = instance.query(
+            "SELECT exceptions.text FROM system.kafka_consumers ARRAY JOIN exceptions WHERE table = 'foo_exceptions'"
+        )
+        logging.debug(f"system.kafka_consumers content: {all_exceptions}")
+        raise AssertionError(
+            f"Expected at least one entry in system.kafka_consumers.exceptions starting with "
+            f"{expected_prefix!r}, but none was found. Captured exceptions:\n{all_exceptions}"
+        )
 
     instance.query("DROP TABLE foo_exceptions")
 

